components/emoji_picker/emoji_picker.jsx

```diff
@@ -38,358 +120,268 @@ export default class EmojiPicker extends React.Component {
+                    return EmojiStore.has(name);
+                }).map((name) => {
+                    return EmojiStore.get(name);
+                });
+            } else {
+                const indices = Emoji.EmojiIndicesByCategory.get(category) || [];
+                categoryEmojis = indices.map((index) => Emoji.Emojis[index]);
+                if (category === 'custom') {
+                    categoryEmojis = categoryEmojis.concat([...EmojiStore.getCustomEmojiMap().values()]);
                 }
             }
-
-            return false;
-        });
-    }
-
-    generateList(filter) {
-        if (filter) {
-            return this.generateFilteredList(filter);
-        }
-        let list = [];
-
-        for (const category of Object.keys(this.categories)) {
-            const emojis = this.getEmojis(category, filter);
-
-            if (emojis.length) {
-                this.categories[category].offset = list.length * ROW_SIZE;
-                this.categories[category].enable = true;
-
-                const emojiHeaderRow = this.generateEmojiHeaderRow(category);
-                list.push(emojiHeaderRow);
-
-                const emojiRows = this.generateEmojiRows(emojis, category);
-                list = this.addEmojiRow(list, emojiRows);
+            categories[category].emojiIds = categoryEmojis.map((emoji) => getEmojiFilename(emoji));
+            for (let i = 0; i < categoryEmojis.length; i++) {
+                const currentEmoji = categoryEmojis[i];
+                const fileName = getEmojiFilename(currentEmoji);
+                allEmojis[fileName] = {
+                    ...currentEmoji,
+                    visible: false,
+                    offset: null
+                };
+                if (!currentEmoji.filename) {
+                    // if custom emoji, set proper attributes
+                    allEmojis[fileName] = {
+                        ...allEmojis[fileName],
+                        aliases: [currentEmoji.name],
+                        category: 'custom',
+                        filename: fileName
+                    };
+                }
             }
         }
-
-        return list;
+        this.setState({
+            categories,
+            allEmojis
+        });
     }
 
-    generateFilteredList(filter) {
-        let emojis = [];
-
-        for (const category of Object.keys(this.categories)) {
-            if (category !== 'recent') {
-                emojis = emojis.concat(this.getEmojis(category, filter));
-            }
-        }
-
-        let list = [];
-        if (emojis.length) {
-            const emojiHeaderRow = this.generateEmojiHeaderRow(CATEGORY_SEARCH_RESULTS);
-            list.push(emojiHeaderRow);
-
-            const emojiRows = this.generateEmojiRows(emojis, CATEGORY_SEARCH_RESULTS);
-            list = this.addEmojiRow(list, emojiRows);
-        }
-
-        return list;
+    getCurrentEmojiCategoryName() {
```

Comment:
Regarding the previous comment, I think we'll need to change this on scroll instead of hover. It may be nicer when you're using the keyboard, but if you use the scroll bar or keep your mouse still while using the scroll wheel, it doesn't update the tabs properly.

Reply:
okay. I will start to update this to reflect prior functionality